faststreams/outputs.nim

```diff
@@ -737,6 +737,19 @@ when fsAsyncSupport:
     writeAndWait(sp, memCopyToBytes(value))
 
 proc writeBytesToCursor(c: var WriteCursor, bytes: openArray[byte]) =
+  # Nim represents a zero-length openArray as a (NULL, 0) base+length tuple.
```

Comment:
Seems a bit weird to have to know Nim internals and C behaviour.

It is not documented that `copyMem` maps to `memcpy/memmove`, and even if it would do so, Nim should simply never emit code that exhibits those problems. So, the proper fix would more likely be to adjust `copyMem` to not call C functions tagged as `_Nonnull` / `assume_not_null_begin` with `null` pointers.

Until that's done, still better to fix it in the callers to avoid random bugs. But a Nim issue link could still be nice here to know when it's fixed and this check can be removed once more.

Reply:
https://github.com/nim-lang/Nim/issues/22149

https://github.com/status-im/nim-faststreams/pull/45/commits/26679784ae6c8beba1df66192f2fd26edf189594

Reply:
Nim does not necessarily represent as a zero-length openArray as a (NULL, 0) base+length tuple. I don't know what it currently does but in the past it didn't because it's stupid.

Reply:
"necessarily" is doing a lot of work. Aside from exactly this representation being part of the trigger for this PR, empirically.

```nim
proc f(a: openArray[int]) = discard

proc g() =
  var foo_var: seq[int]
  let foo_let: seq[int] = @[]
  const foo_const: seq[int] = @[]
  doAssert len(foo_var) == 0
  doAssert len(foo_let) == 0
  doAssert len(foo_const) == 0
  f(foo_var)
  f(foo_let)
  f(foo_const)

g()
```
compiles into
```c
N_LIB_PRIVATE N_NIMCALL(void, f__c_u1)(NI* a_p0, NI a_p0Len_0) {
	nimfr_("f", "/home/user/c.nim");
	popFrame();
}

...

static N_INLINE(NIM_BOOL*, nimErrorFlag)(void) {
	NIM_BOOL* result;
	result = (NIM_BOOL*)0;
	result = (&nimInErrorMode__system_u4393);
	return result;
}
N_LIB_PRIVATE N_NIMCALL(void, g__c_u3)(void) {
	tySequence__qwqHTkRvwhrRyENtudHQ7g foo_var;
	tySequence__qwqHTkRvwhrRyENtudHQ7g foo_let;
	tySequence__qwqHTkRvwhrRyENtudHQ7g T15_;
NIM_BOOL* nimErr_;
	nimfr_("g", "/home/user/c.nim");
{nimErr_ = nimErrorFlag();
	foo_var.len = 0; foo_var.p = NIM_NIL;
	foo_let.len = 0; foo_let.p = NIM_NIL;
	nimln_(5);
	foo_let.len = 0; foo_let.p = (tySequence__qwqHTkRvwhrRyENtudHQ7g_Content*) newSeqPayload(0, sizeof(NI), NIM_ALIGNOF(NI));
	nimln_(7);
	{
		NI T3_;
		T3_ = foo_var.len;
		if (!!((T3_ == ((NI)0)))) goto LA4_;
		failedAssertImpl__stdZassertions_u315(TM__PWLVblvdk3wLpEhn9cPvLyw_3);
		if (NIM_UNLIKELY(*nimErr_)) goto BeforeRet_;
	}
LA4_: ;
	nimln_(8);
	{
		NI T8_;
		T8_ = foo_let.len;
		if (!!((T8_ == ((NI)0)))) goto LA9_;
		failedAssertImpl__stdZassertions_u315(TM__PWLVblvdk3wLpEhn9cPvLyw_5);
		if (NIM_UNLIKELY(*nimErr_)) goto BeforeRet_;
	}
LA9_: ;
	nimln_(9);
	{
		if (!NIM_FALSE) goto LA13_;
		failedAssertImpl__stdZassertions_u315(TM__PWLVblvdk3wLpEhn9cPvLyw_7);
		if (NIM_UNLIKELY(*nimErr_)) goto BeforeRet_;
	}
LA13_: ;
	nimln_(10);
	f__c_u1(((foo_var).p) ? (foo_var.p->data) : NIM_NIL, foo_var.len);
	if (NIM_UNLIKELY(*nimErr_)) goto BeforeRet_;
	nimln_(11);
	f__c_u1(((foo_let).p) ? (foo_let.p->data) : NIM_NIL, foo_let.len);
	if (NIM_UNLIKELY(*nimErr_)) goto BeforeRet_;
	nimln_(12);
	T15_.len = 0; T15_.p = NIM_NIL;
	T15_.len = 0; T15_.p = (tySequence__qwqHTkRvwhrRyENtudHQ7g_Content*) newSeqPayload(0, sizeof(NI), NIM_ALIGNOF(NI));
	f__c_u1(((T15_).p) ? (T15_.p->data) : NIM_NIL, T15_.len);
	if (NIM_UNLIKELY(*nimErr_)) goto BeforeRet_;
	nimln_(4);
	eqdestroy___c_u43((&foo_let));
	}BeforeRet_: ;
	popFrame();
}
```

where `newSeqPayload` (only called for `let`, not `var` amounts to a no-op here:
```c
N_LIB_PRIVATE N_NIMCALL(void*, newSeqPayload)(NI cap_p0, NI elemSize_p1, NI elemAlign_p2) {
	void* result;
{	result = (void*)0;
	{
		tyObject_NimSeqPayloadBase__uNU0QIKyz59aK9bu2urUuQ1Q* p;
		NI T5_;
		NI TM__Q5wkpxktOdTGvlSRo9bzt9aw_49;
		NI TM__Q5wkpxktOdTGvlSRo9bzt9aw_50;
		void* T6_;
		if (!(((NI)0) < cap_p0)) goto LA3_;
		T5_ = (NI)0;
		T5_ = align__system_u1628(((NI)8), elemAlign_p2);
		if (nimMulInt(cap_p0, elemSize_p1, &TM__Q5wkpxktOdTGvlSRo9bzt9aw_49)) { raiseOverflow(); goto BeforeRet_;
		};
		if (nimAddInt(T5_, (NI)(TM__Q5wkpxktOdTGvlSRo9bzt9aw_49), &TM__Q5wkpxktOdTGvlSRo9bzt9aw_50)) { raiseOverflow(); goto BeforeRet_;
		};
		if (((NI)(TM__Q5wkpxktOdTGvlSRo9bzt9aw_50)) < ((NI)0) || ((NI)(TM__Q5wkpxktOdTGvlSRo9bzt9aw_50)) > ((NI)IL64(9223372036854775807))){ raiseRangeErrorI((NI)(TM__Q5wkpxktOdTGvlSRo9bzt9aw_50), ((NI)0), ((NI)IL64(9223372036854775807))); goto BeforeRet_;
		}
		if ((elemAlign_p2) < ((NI)0) || (elemAlign_p2) > ((NI)IL64(9223372036854775807))){ raiseRangeErrorI(elemAlign_p2, ((NI)0), ((NI)IL64(9223372036854775807))); goto BeforeRet_;
		}
		T6_ = (void*)0;
		T6_ = alignedAlloc0__system_u1939(((NI) ((NI)(TM__Q5wkpxktOdTGvlSRo9bzt9aw_50))), ((NI) (elemAlign_p2)));
		p = ((tyObject_NimSeqPayloadBase__uNU0QIKyz59aK9bu2urUuQ1Q*) (T6_));
		(*p).cap = cap_p0;
		result = ((void*) (p));
	}
	goto LA1_;
LA3_: ;
	{
		result = NIM_NIL;
	}
LA1_: ;
	}BeforeRet_: ;
	return result;
}
```
because `cap_p0` is 0, and therefore `if (!(((NI)0) < cap_p0)) goto LA3_;` triggers an early exit.

This is all in current `devel`, using default ORC, but the general behavior has been this way for a while.

Reply:
Well I was against `(foo_var).p) ? (foo_var.p->data) : NIM_NIL` special casing because it's silly. *shrug*

Reply:
`(foo_var).p) ? (foo_var.p->data) : NIM_NIL` and similar constructions are downstream consequences, not upstream causes. Once the, yes, empty `seq`/`openArray` having `NIM_NIL` address convention was established, doing anything else becomes UB.

Reply:
Incidentally, the same `NULL`-by-default problem occurs with empty strings: https://github.com/nim-lang/Nim/issues/22153

And the correct answer isn't to just ignore it there, either.

Reply:
The correct answer here is to report these things as bugs to the sanitizers and the C spec maintainers. It's clearly 100% against the spirit of C and does not reflect how address computations work on any CPU ever invented.

Reply:
>  It's clearly 100% against the spirit of C and does not reflect how address computations work on any CPU ever invented.

I'm willing to bet that the C spec maintainers were quite tune with the C language as they were writing the spec since they took pains to explicitly point out that pointers must be non-null even when length is zero - among other things, it simplifies the loop when copying the full range of the platform `int`, a point you surely can appreciate from all the bug reports and PR:s pointing out how silly it can be not to be able to iterate to `uint64.high` in a language.

You may also be interested in how compilers reason about this - gcc for example explicitly states that it can remove `NULL` checks based on `mem*` argument analysis: https://gcc.gnu.org/gcc-4.9/porting_to.html (see "null pointer checks...").

It is of course possible that _both_ the C spec maintainers and the compiler writers were out of tune with the language they were working with, but... :)

Reply:
That's just a long winded way of stating an argument by authority.

Reply:
How do you mean, authority? the gcc optimization _and_ the specific way of optimising memcpy are both ways in which the UB can turn into actual problems for the user - ie Nim is clearly in the wrong here - you may disagree with the spirit and idea (by your authority) but this doesn't mean that the code is well-formed according to the specification / standard - now, if we want to make Nim have fewer cases of undefined behavior than C, the right thing to do here is simply to include a length check in `copyMem`/etc _or_ at least an assert that gets disabled on `-d:danger`.

Reply:
I don't mind a length check in copyMem but the patch that introduced `(foo_var).p) ? (foo_var.p->data) : NIM_NIL` might introduce more problems than it solved. Also, assuming somebody gets fixes into the C spec eventually, we can remove these hacks again.